Extend Inductor to support the third-party backend

EikanWang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #100706

This PR intends to extend Inductor to support the third-party backend that only focuses on the code generation just like what C++/OpenMP and Triton backend have done. 

Currently, the generated code by Inductor contains two major parts. One is the kernel, and the other is the Python wrapper to glue the kernel. Therefore, the third-party backend needs to customize the two parts to generate its specific code.

- Python wrapper code generation

  Inductor provides a `WrapperCodeGen` class to generate the Python wrapper code to glue the kernel. Therefore, it is straightforward for the third-party backend to generate the backend-specific Python wrapper code. It just needs to inherit the `WrapperCodeGen` class and purposely override the particular member functions. 

- Kernel code generation
  
  It is driven by different `Scheduling`. Hence, the third-party backend needs to provide a custom `Scheduling` for its specific kernel code generation. Currently, `CppScheduling` and `TritonScheduling` are for C++/OpenMP and Triton backend, respectively. But there is no common `Scheduling` class. Based on the scheduling invocation, this PR abstracts a common `Scheduling` class containing the following member functions. 

  -   [group_fn](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#LL649C64-L649C64)
  - [flush](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#L1150)
  - [can_fuse_vertical](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#L1006)
  - [can_fuse_horizontal](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#LL1008C45-L1008C64)
  - [codegen_template](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#L1234) _This function is only available for triton. If the third-party backend behaves as a sub-class of `TritonScheduling`, it can override it or reuse it._
  - [codegen_nodes](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#L1234)
  - [codegen_sync](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#LL1251C1-L1251C1). _This function is only available for triton debug purpose. But it might also be useful for other computation devices. Therefore, we'd prefer to keep this function._ 
  
  The third-party backend needs to inherit from the `Scheduling` class and implement these functions.


Regarding some other classes like `CppKernel` and `TritonKernel` for code generation, they are used by or part of the logic of either `Scheduling` or `WrapperCodeGen`. Hence, this PR does not define the interface and leaves the flexibility to the third-party backend. The third-party backend can decide to implement these classes from scratch or reuse them by inheriting and overriding them.

cc @voznesenskym @penguinwu @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @soumith @desertfire

EikanWang committed:
Support the third-party backend that could implement in PyTorch extension for different devices

[ghstack-poisoned]

EikanWang committed:
Update on "Support the third-party backend that could implement in PyTorch extension for different devices"




cc soumith voznesenskym penguinwu jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang committed:
Update on "Support the third-party backend that could implement in PyTorch extension for different devices"




cc soumith voznesenskym penguinwu jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang committed:
Update on "Support the third-party backend that could implement in PyTorch extension for different devices"




cc soumith voznesenskym penguinwu jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang committed:
Update on "Support the third-party backend that could implement in PyTorch extension for different devices"


This PR intends to extend Inductor to support the third-party backend that only focuses on the code generation just like what C++/OpenMP and Triton backend have done. 

Currently, the generated code by Inductor contains two major parts. One is the kernel, and the other is the Python wrapper to glue the kernel. Therefore, the third-party backend needs to customize the two parts to generate its specific code.

- Python wrapper code generation

  Inductor provides a `WrapperCodeGen` class to generate the Python wrapper code to glue the kernel. Therefore, it is straightforward for the third-party backend to generate the backend-specific Python wrapper code. It just needs to inherit the `WrapperCodeGen` class and purposely override the particular member functions. 

- Kernel code generation
  
  It is driven by different `Scheduling`. Hence, the third-party backend needs to provide a custom `Scheduling` for its specific kernel code generation. Currently, `CppScheduling` and `TritonScheduling` are for C++/OpenMP and Triton backend, respectively. But there is no common `Scheduling` class. Based on the scheduling invocation, this PR abstracts a common `Scheduling` class containing the following member functions. 

  -   [group_fn](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#LL649C64-L649C64)
  - [flush](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#L1150)
  - [can_fuse_vertical](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#L1006)
  - [can_fuse_horizontal](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#LL1008C45-L1008C64)
  - [codegen_template](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#L1234) _This function is only available for triton. If the third-party backend behaves as a sub-class of `TritonScheduling`, it can override it or reuse it._
  - [codegen_nodes](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#L1234)
  - [codegen_sync](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#LL1251C1-L1251C1). _This function is only available for triton debug purpose. But it might also be useful for other computation devices. Therefore, we'd prefer to keep this function._ 
  
  The third-party backend needs to inherit from the `Scheduling` class and implement these functions.


Regarding some other classes like `CppKernel` and `TritonKernel` for code generation, they are used by or part of the logic of either `Scheduling` or `WrapperCodeGen`. Hence, this PR does not define the interface and leaves the flexibility to the third-party backend. The third-party backend can decide to implement these classes from scratch or reuse them by inheriting and overriding them.

cc soumith voznesenskym penguinwu jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jgong5 commented on torch/_inductor/graph.py:
```diff
@@ -633,7 +634,18 @@ def init_wrapper_code(self):
 
-        self.wrapper_code = WrapperCodeGen()
+        if not all(device_type in ["cuda", "cpu"] for device_type in self.device_types):
+            device_types_copy = self.device_types
+            device_types_copy.discard("cuda")
+            device_types_copy.discard("cpu")
```
this changes `self.device_types` inplace?

jgong5 commented on torch/_inductor/utils.py:
```diff
@@ -1075,3 +1078,18 @@ def triton_config_to_hashable(cfg):
     return tuple(items)
+
+
+def register_backend_for_device(
```
better to make it a decorator on custom `Scheduling` and `WrapperCodegen"?

jgong5 commented on torch/_inductor/codegen/scheduling.py:
```diff
@@ -0,0 +1,31 @@
+        pass
+
+    @abstractmethod
+    def codegen_nodes(self, *args, **kwargs):
```
why it is not a list of scheduling nodes?

jgong5 commented on torch/_inductor/codegen/scheduling.py:
```diff
@@ -0,0 +1,31 @@
+        pass
+
+    @abstractmethod
+    def codegen_template(self, *args, **kwargs):
```
why is it not template node and epilogues?

jansel replied:
I don't think this will work with non-triton backends currently.

jgong5 commented on torch/_inductor/codegen/scheduling.py:
better to type the methods in `Scheduling` and avoid using `*args` and `**kwargs` if possible.

jgong5 commented on torch/_inductor/graph.py:
```diff
@@ -633,7 +634,18 @@ def init_wrapper_code(self):
+                "+".join(device_types_copy)
+            )
+            wrapper_code_gen_cls = get_wrapper_for_device(device_types_copy.pop())
+            assert wrapper_code_gen_cls
```
add error message here.

jansel commented on torch/_inductor/utils.py:
```diff
@@ -31,6 +31,9 @@
 
 VarRanges = Dict[sympy.Expr, sympy.Expr]
 
+extension_scheduling: Dict[str, type] = {}
```
utils.py seems like the wrong place for this registration stuff.  Either a new file or scheduler.py.

jansel commented on torch/_inductor/scheduler.py:
```diff
@@ -1184,6 +1190,8 @@ def create_backend(self, device: torch.device):
             from .codegen.triton import TritonScheduling
 
             return TritonScheduling(self)
+        elif get_scheduling_for_device(device.type):
```
We should refactor the lines above to use the same registration system.

jansel commented on torch/_inductor/codegen/scheduling.py:
```diff
@@ -0,0 +1,31 @@
+        pass
+
+    @abstractmethod
+    def codegen_sync(self, *args, **kwargs):
```
Use actual function signatures not args/kwargs.

jansel commented on torch/_inductor/codegen/scheduling.py:
```diff
@@ -0,0 +1,31 @@
+from abc import ABC, abstractmethod
+
+
+class Scheduling(ABC):
```
We should modify the existing backends to inherit from this. 

Add docstrings for every method.

Perhaps this belongs in common.py, or some sort of backend registration file.

jansel commented on test/inductor/extension_backends/extension_scheduling.py:
```diff
@@ -0,0 +1,37 @@
+from torch._inductor.codegen import cpp, wrapper
+from torch._inductor.codegen.scheduling import Scheduling
+from torch._inductor.virtualized import V
+
+
+class ExtensionWrapperCodegen(wrapper.WrapperCodeGen):
+    def __init__(self):
+        super().__init__()
+
+
+class ExtensionScheduling(Scheduling):
+    def __init__(self, scheduler):
+        self.scheduler = scheduler
+        self._scheduling = cpp.CppScheduling(scheduler)
+
+    def can_fuse_vertical(self, node1, node2):
+        return True
+
+    def can_fuse_horizontal(self, node1, node2):
+        return True
+
+    def group_fn(self, *args, **kwargs):
+        sizes = args[0]
+        return tuple(tuple(map(V.graph.sizevars.simplify, s)) for s in sizes)
+
+    def codegen_template(self, *args, **kwargs):
+        pass
+
+    def codegen_nodes(self, *args, **kwargs):
+        nodes = args[0]
+        self._scheduling.codegen_nodes(nodes)
+
+    def codegen_sync(self, *args, **kwargs):
+        pass
+
+    def flush(self, *args, **kwargs):
+        self._scheduling.flush()
```
I was hoping to see a more complete example than this.  One that actually generates some minimal working C++ code.  I worry there are hidden issues that will be uncovered.

EikanWang replied:
Do you think partial elementwise operations are good enough? Or should we implement all the functions?

jansel replied:
It doesn't need to implement everything, but it should demonstrate the key concepts:
1) pointwise ops
2) reduction ops
3) horizontal fusion
4) vertical fusion

EikanWang replied:
Okay. I will develop a minimal codegen to demonstrate these key concepts.

EikanWang committed:
Update on "Extend Inductor to support the third-party backend"


This PR intends to extend Inductor to support the third-party backend that only focuses on the code generation just like what C++/OpenMP and Triton backend have done. 

Currently, the generated code by Inductor contains two major parts. One is the kernel, and the other is the Python wrapper to glue the kernel. Therefore, the third-party backend needs to customize the two parts to generate its specific code.

- Python wrapper code generation

  Inductor provides a `WrapperCodeGen` class to generate the Python wrapper code to glue the kernel. Therefore, it is straightforward for the third-party backend to generate the backend-specific Python wrapper code. It just needs to inherit the `WrapperCodeGen` class and purposely override the particular member functions. 

- Kernel code generation
  
  It is driven by different `Scheduling`. Hence, the third-party backend needs to provide a custom `Scheduling` for its specific kernel code generation. Currently, `CppScheduling` and `TritonScheduling` are for C++/OpenMP and Triton backend, respectively. But there is no common `Scheduling` class. Based on the scheduling invocation, this PR abstracts a common `Scheduling` class containing the following member functions. 

  -   [group_fn](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#LL649C64-L649C64)
  - [flush](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#L1150)
  - [can_fuse_vertical](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#L1006)
  - [can_fuse_horizontal](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#LL1008C45-L1008C64)
  - [codegen_template](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#L1234) _This function is only available for triton. If the third-party backend behaves as a sub-class of `TritonScheduling`, it can override it or reuse it._
  - [codegen_nodes](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#L1234)
  - [codegen_sync](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#LL1251C1-L1251C1). _This function is only available for triton debug purpose. But it might also be useful for other computation devices. Therefore, we'd prefer to keep this function._ 
  
  The third-party backend needs to inherit from the `Scheduling` class and implement these functions.


Regarding some other classes like `CppKernel` and `TritonKernel` for code generation, they are used by or part of the logic of either `Scheduling` or `WrapperCodeGen`. Hence, this PR does not define the interface and leaves the flexibility to the third-party backend. The third-party backend can decide to implement these classes from scratch or reuse them by inheriting and overriding them.

cc soumith voznesenskym penguinwu jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang committed:
Update on "Extend Inductor to support the third-party backend"


This PR intends to extend Inductor to support the third-party backend that only focuses on the code generation just like what C++/OpenMP and Triton backend have done. 

Currently, the generated code by Inductor contains two major parts. One is the kernel, and the other is the Python wrapper to glue the kernel. Therefore, the third-party backend needs to customize the two parts to generate its specific code.

- Python wrapper code generation

  Inductor provides a `WrapperCodeGen` class to generate the Python wrapper code to glue the kernel. Therefore, it is straightforward for the third-party backend to generate the backend-specific Python wrapper code. It just needs to inherit the `WrapperCodeGen` class and purposely override the particular member functions. 

- Kernel code generation
  
  It is driven by different `Scheduling`. Hence, the third-party backend needs to provide a custom `Scheduling` for its specific kernel code generation. Currently, `CppScheduling` and `TritonScheduling` are for C++/OpenMP and Triton backend, respectively. But there is no common `Scheduling` class. Based on the scheduling invocation, this PR abstracts a common `Scheduling` class containing the following member functions. 

  -   [group_fn](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#LL649C64-L649C64)
  - [flush](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#L1150)
  - [can_fuse_vertical](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#L1006)
  - [can_fuse_horizontal](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#LL1008C45-L1008C64)
  - [codegen_template](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#L1234) _This function is only available for triton. If the third-party backend behaves as a sub-class of `TritonScheduling`, it can override it or reuse it._
  - [codegen_nodes](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#L1234)
  - [codegen_sync](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#LL1251C1-L1251C1). _This function is only available for triton debug purpose. But it might also be useful for other computation devices. Therefore, we'd prefer to keep this function._ 
  
  The third-party backend needs to inherit from the `Scheduling` class and implement these functions.


Regarding some other classes like `CppKernel` and `TritonKernel` for code generation, they are used by or part of the logic of either `Scheduling` or `WrapperCodeGen`. Hence, this PR does not define the interface and leaves the flexibility to the third-party backend. The third-party backend can decide to implement these classes from scratch or reuse them by inheriting and overriding them.

cc soumith voznesenskym penguinwu jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang committed:
Update on "Extend Inductor to support the third-party backend"


This PR intends to extend Inductor to support the third-party backend that only focuses on the code generation just like what C++/OpenMP and Triton backend have done. 

Currently, the generated code by Inductor contains two major parts. One is the kernel, and the other is the Python wrapper to glue the kernel. Therefore, the third-party backend needs to customize the two parts to generate its specific code.

- Python wrapper code generation

  Inductor provides a `WrapperCodeGen` class to generate the Python wrapper code to glue the kernel. Therefore, it is straightforward for the third-party backend to generate the backend-specific Python wrapper code. It just needs to inherit the `WrapperCodeGen` class and purposely override the particular member functions. 

- Kernel code generation
  
  It is driven by different `Scheduling`. Hence, the third-party backend needs to provide a custom `Scheduling` for its specific kernel code generation. Currently, `CppScheduling` and `TritonScheduling` are for C++/OpenMP and Triton backend, respectively. But there is no common `Scheduling` class. Based on the scheduling invocation, this PR abstracts a common `Scheduling` class containing the following member functions. 

  -   [group_fn](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#LL649C64-L649C64)
  - [flush](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#L1150)
  - [can_fuse_vertical](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#L1006)
  - [can_fuse_horizontal](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#LL1008C45-L1008C64)
  - [codegen_template](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#L1234) _This function is only available for triton. If the third-party backend behaves as a sub-class of `TritonScheduling`, it can override it or reuse it._
  - [codegen_nodes](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#L1234)
  - [codegen_sync](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#LL1251C1-L1251C1). _This function is only available for triton debug purpose. But it might also be useful for other computation devices. Therefore, we'd prefer to keep this function._ 
  
  The third-party backend needs to inherit from the `Scheduling` class and implement these functions.


Regarding some other classes like `CppKernel` and `TritonKernel` for code generation, they are used by or part of the logic of either `Scheduling` or `WrapperCodeGen`. Hence, this PR does not define the interface and leaves the flexibility to the third-party backend. The third-party backend can decide to implement these classes from scratch or reuse them by inheriting and overriding them.

cc soumith voznesenskym penguinwu jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jgong5 commented:
BTW, can we also support overriding the backend of registered devices?

EikanWang committed:
Update on "Extend Inductor to support the third-party backend"


This PR intends to extend Inductor to support the third-party backend that only focuses on the code generation just like what C++/OpenMP and Triton backend have done. 

Currently, the generated code by Inductor contains two major parts. One is the kernel, and the other is the Python wrapper to glue the kernel. Therefore, the third-party backend needs to customize the two parts to generate its specific code.

- Python wrapper code generation

  Inductor provides a `WrapperCodeGen` class to generate the Python wrapper code to glue the kernel. Therefore, it is straightforward for the third-party backend to generate the backend-specific Python wrapper code. It just needs to inherit the `WrapperCodeGen` class and purposely override the particular member functions. 

- Kernel code generation
  
  It is driven by different `Scheduling`. Hence, the third-party backend needs to provide a custom `Scheduling` for its specific kernel code generation. Currently, `CppScheduling` and `TritonScheduling` are for C++/OpenMP and Triton backend, respectively. But there is no common `Scheduling` class. Based on the scheduling invocation, this PR abstracts a common `Scheduling` class containing the following member functions. 

  -   [group_fn](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#LL649C64-L649C64)
  - [flush](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#L1150)
  - [can_fuse_vertical](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#L1006)
  - [can_fuse_horizontal](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#LL1008C45-L1008C64)
  - [codegen_template](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#L1234) _This function is only available for triton. If the third-party backend behaves as a sub-class of `TritonScheduling`, it can override it or reuse it._
  - [codegen_nodes](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#L1234)
  - [codegen_sync](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#LL1251C1-L1251C1). _This function is only available for triton debug purpose. But it might also be useful for other computation devices. Therefore, we'd prefer to keep this function._ 
  
  The third-party backend needs to inherit from the `Scheduling` class and implement these functions.


Regarding some other classes like `CppKernel` and `TritonKernel` for code generation, they are used by or part of the logic of either `Scheduling` or `WrapperCodeGen`. Hence, this PR does not define the interface and leaves the flexibility to the third-party backend. The third-party backend can decide to implement these classes from scratch or reuse them by inheriting and overriding them.

cc soumith voznesenskym penguinwu jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang commented:
> BTW, can we also support overriding the backend of registered devices?

In practice, this PR should have supported this scenario.
https://github.com/pytorch/pytorch/pull/100706/files#diff-e7f291bf7d847fea77a46bd4c0ce6ee0eacf851714a397b4271e54f36fedf369R672

EikanWang commented:
Just update the status here. We are working on a custom Trtion backend for Intel GPU.

EikanWang committed:
Update on "Extend Inductor to support the third-party backend"


This PR intends to extend Inductor to support the third-party backend that only focuses on the code generation just like what C++/OpenMP and Triton backend have done. 

Currently, the generated code by Inductor contains two major parts. One is the kernel, and the other is the Python wrapper to glue the kernel. Therefore, the third-party backend needs to customize the two parts to generate its specific code.

- Python wrapper code generation

  Inductor provides a `WrapperCodeGen` class to generate the Python wrapper code to glue the kernel. Therefore, it is straightforward for the third-party backend to generate the backend-specific Python wrapper code. It just needs to inherit the `WrapperCodeGen` class and purposely override the particular member functions. 

- Kernel code generation
  
  It is driven by different `Scheduling`. Hence, the third-party backend needs to provide a custom `Scheduling` for its specific kernel code generation. Currently, `CppScheduling` and `TritonScheduling` are for C++/OpenMP and Triton backend, respectively. But there is no common `Scheduling` class. Based on the scheduling invocation, this PR abstracts a common `Scheduling` class containing the following member functions. 

  -   [group_fn](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#LL649C64-L649C64)
  - [flush](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#L1150)
  - [can_fuse_vertical](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#L1006)
  - [can_fuse_horizontal](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#LL1008C45-L1008C64)
  - [codegen_template](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#L1234) _This function is only available for triton. If the third-party backend behaves as a sub-class of `TritonScheduling`, it can override it or reuse it._
  - [codegen_nodes](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#L1234)
  - [codegen_sync](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#LL1251C1-L1251C1). _This function is only available for triton debug purpose. But it might also be useful for other computation devices. Therefore, we'd prefer to keep this function._ 
  
  The third-party backend needs to inherit from the `Scheduling` class and implement these functions.


Regarding some other classes like `CppKernel` and `TritonKernel` for code generation, they are used by or part of the logic of either `Scheduling` or `WrapperCodeGen`. Hence, this PR does not define the interface and leaves the flexibility to the third-party backend. The third-party backend can decide to implement these classes from scratch or reuse them by inheriting and overriding them.

cc soumith voznesenskym penguinwu jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang committed:
Update on "Extend Inductor to support the third-party backend"


This PR intends to extend Inductor to support the third-party backend that only focuses on the code generation just like what C++/OpenMP and Triton backend have done. 

Currently, the generated code by Inductor contains two major parts. One is the kernel, and the other is the Python wrapper to glue the kernel. Therefore, the third-party backend needs to customize the two parts to generate its specific code.

- Python wrapper code generation

  Inductor provides a `WrapperCodeGen` class to generate the Python wrapper code to glue the kernel. Therefore, it is straightforward for the third-party backend to generate the backend-specific Python wrapper code. It just needs to inherit the `WrapperCodeGen` class and purposely override the particular member functions. 

- Kernel code generation
  
  It is driven by different `Scheduling`. Hence, the third-party backend needs to provide a custom `Scheduling` for its specific kernel code generation. Currently, `CppScheduling` and `TritonScheduling` are for C++/OpenMP and Triton backend, respectively. But there is no common `Scheduling` class. Based on the scheduling invocation, this PR abstracts a common `Scheduling` class containing the following member functions. 

  -   [group_fn](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#LL649C64-L649C64)
  - [flush](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#L1150)
  - [can_fuse_vertical](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#L1006)
  - [can_fuse_horizontal](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#LL1008C45-L1008C64)
  - [codegen_template](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#L1234) _This function is only available for triton. If the third-party backend behaves as a sub-class of `TritonScheduling`, it can override it or reuse it._
  - [codegen_nodes](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#L1234)
  - [codegen_sync](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#LL1251C1-L1251C1). _This function is only available for triton debug purpose. But it might also be useful for other computation devices. Therefore, we'd prefer to keep this function._ 
  
  The third-party backend needs to inherit from the `Scheduling` class and implement these functions.


Regarding some other classes like `CppKernel` and `TritonKernel` for code generation, they are used by or part of the logic of either `Scheduling` or `WrapperCodeGen`. Hence, this PR does not define the interface and leaves the flexibility to the third-party backend. The third-party backend can decide to implement these classes from scratch or reuse them by inheriting and overriding them.

cc soumith voznesenskym penguinwu jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang committed:
Update on "Extend Inductor to support the third-party backend"


This PR intends to extend Inductor to support the third-party backend that only focuses on the code generation just like what C++/OpenMP and Triton backend have done. 

Currently, the generated code by Inductor contains two major parts. One is the kernel, and the other is the Python wrapper to glue the kernel. Therefore, the third-party backend needs to customize the two parts to generate its specific code.

- Python wrapper code generation

  Inductor provides a `WrapperCodeGen` class to generate the Python wrapper code to glue the kernel. Therefore, it is straightforward for the third-party backend to generate the backend-specific Python wrapper code. It just needs to inherit the `WrapperCodeGen` class and purposely override the particular member functions. 

- Kernel code generation
  
  It is driven by different `Scheduling`. Hence, the third-party backend needs to provide a custom `Scheduling` for its specific kernel code generation. Currently, `CppScheduling` and `TritonScheduling` are for C++/OpenMP and Triton backend, respectively. But there is no common `Scheduling` class. Based on the scheduling invocation, this PR abstracts a common `Scheduling` class containing the following member functions. 

  -   [group_fn](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#LL649C64-L649C64)
  - [flush](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#L1150)
  - [can_fuse_vertical](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#L1006)
  - [can_fuse_horizontal](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#LL1008C45-L1008C64)
  - [codegen_template](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#L1234) _This function is only available for triton. If the third-party backend behaves as a sub-class of `TritonScheduling`, it can override it or reuse it._
  - [codegen_nodes](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#L1234)
  - [codegen_sync](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#LL1251C1-L1251C1). _This function is only available for triton debug purpose. But it might also be useful for other computation devices. Therefore, we'd prefer to keep this function._ 
  
  The third-party backend needs to inherit from the `Scheduling` class and implement these functions.


Regarding some other classes like `CppKernel` and `TritonKernel` for code generation, they are used by or part of the logic of either `Scheduling` or `WrapperCodeGen`. Hence, this PR does not define the interface and leaves the flexibility to the third-party backend. The third-party backend can decide to implement these classes from scratch or reuse them by inheriting and overriding them.

cc soumith voznesenskym penguinwu jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang committed:
Update on "Extend Inductor to support the third-party backend"


This PR intends to extend Inductor to support the third-party backend that only focuses on the code generation just like what C++/OpenMP and Triton backend have done. 

Currently, the generated code by Inductor contains two major parts. One is the kernel, and the other is the Python wrapper to glue the kernel. Therefore, the third-party backend needs to customize the two parts to generate its specific code.

- Python wrapper code generation

  Inductor provides a `WrapperCodeGen` class to generate the Python wrapper code to glue the kernel. Therefore, it is straightforward for the third-party backend to generate the backend-specific Python wrapper code. It just needs to inherit the `WrapperCodeGen` class and purposely override the particular member functions. 

- Kernel code generation
  
  It is driven by different `Scheduling`. Hence, the third-party backend needs to provide a custom `Scheduling` for its specific kernel code generation. Currently, `CppScheduling` and `TritonScheduling` are for C++/OpenMP and Triton backend, respectively. But there is no common `Scheduling` class. Based on the scheduling invocation, this PR abstracts a common `Scheduling` class containing the following member functions. 

  -   [group_fn](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#LL649C64-L649C64)
  - [flush](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#L1150)
  - [can_fuse_vertical](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#L1006)
  - [can_fuse_horizontal](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#LL1008C45-L1008C64)
  - [codegen_template](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#L1234) _This function is only available for triton. If the third-party backend behaves as a sub-class of `TritonScheduling`, it can override it or reuse it._
  - [codegen_nodes](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#L1234)
  - [codegen_sync](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#LL1251C1-L1251C1). _This function is only available for triton debug purpose. But it might also be useful for other computation devices. Therefore, we'd prefer to keep this function._ 
  
  The third-party backend needs to inherit from the `Scheduling` class and implement these functions.


Regarding some other classes like `CppKernel` and `TritonKernel` for code generation, they are used by or part of the logic of either `Scheduling` or `WrapperCodeGen`. Hence, this PR does not define the interface and leaves the flexibility to the third-party backend. The third-party backend can decide to implement these classes from scratch or reuse them by inheriting and overriding them.

cc soumith voznesenskym penguinwu jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang commented on torch/_inductor/codegen/common.py:
```diff
@@ -38,6 +38,46 @@ def data_type_logger(msg):
+device_codegens: typing.Dict[str, DeviceCodegen] = {}
+
+
+# The code generated by Inductor consists of two main parts: kernel code and wrapper code.
+# For any new backend looking to integrate with Inductor, customization of these two main
+# parts are necessary to generate its specific code.
+#
+# Kernel code generation is determined by different Scheduling. Consequently, a new
+# backend needs to provide a custom Scheduling for its unique kernel code generation. Currently,
+# CppScheduling and TritonScheduling serve the C++/OpenMP and Triton backends, respectively.
+#
+# For the Wrapper, Inductor provides a WrapperCodeGen class to generate the Python wrapper code
+# that bridges kernels. This allows out-of-tree backends to inherit from WrapperCodeGen,
+# and override specific member functions to create backend-specific Python wrapper code.
+#
+# Other classes, such as CppKernel and TritonKernel, used for code generation, typically form part
+# of the logic for either Scheduling or WrapperCodeGen. So the Scheduling and WrapperCodeGen interfaces
+# provide flexibility to the backend. A backend can choose to implement these classes from scratch,
+# or reuse them by extending and overriding as necessary. And Inductor provides the registration API,
+# register_backend_for_device, to equip a new backend at runtime.
+#
+# Intel has developed a new backend on top of Triton to support Intel GPUs, leveraging these interfaces.
+# This backend can be used as a reference:
+# https://github.com/intel/intel-extension-for-pytorch/blob/5dcc9d57e5422cf295e1a1ee97896d6b6a554a85/intel_extension_for_pytorch/_inductor/__init__.py#L9
```
@jansel , I added some comments here and provided the example code of IPEX.

EikanWang committed:
Update on "Extend Inductor to support the third-party backend"


This PR intends to extend Inductor to support the third-party backend that only focuses on the code generation just like what C++/OpenMP and Triton backend have done. 

Currently, the generated code by Inductor contains two major parts. One is the kernel, and the other is the Python wrapper to glue the kernel. Therefore, the third-party backend needs to customize the two parts to generate its specific code.

- Python wrapper code generation

  Inductor provides a `WrapperCodeGen` class to generate the Python wrapper code to glue the kernel. Therefore, it is straightforward for the third-party backend to generate the backend-specific Python wrapper code. It just needs to inherit the `WrapperCodeGen` class and purposely override the particular member functions. 

- Kernel code generation
  
  It is driven by different `Scheduling`. Hence, the third-party backend needs to provide a custom `Scheduling` for its specific kernel code generation. Currently, `CppScheduling` and `TritonScheduling` are for C++/OpenMP and Triton backend, respectively. But there is no common `Scheduling` class. Based on the scheduling invocation, this PR abstracts a common `Scheduling` class containing the following member functions. 

  -   [group_fn](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#LL649C64-L649C64)
  - [flush](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#L1150)
  - [can_fuse_vertical](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#L1006)
  - [can_fuse_horizontal](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#LL1008C45-L1008C64)
  - [codegen_template](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#L1234) _This function is only available for triton. If the third-party backend behaves as a sub-class of `TritonScheduling`, it can override it or reuse it._
  - [codegen_nodes](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#L1234)
  - [codegen_sync](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#LL1251C1-L1251C1). _This function is only available for triton debug purpose. But it might also be useful for other computation devices. Therefore, we'd prefer to keep this function._ 
  
  The third-party backend needs to inherit from the `Scheduling` class and implement these functions.


Regarding some other classes like `CppKernel` and `TritonKernel` for code generation, they are used by or part of the logic of either `Scheduling` or `WrapperCodeGen`. Hence, this PR does not define the interface and leaves the flexibility to the third-party backend. The third-party backend can decide to implement these classes from scratch or reuse them by inheriting and overriding them.

cc soumith voznesenskym penguinwu jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang commented:
@pytorchbot merge

yanboliang commented:
@pytorchbot revert

yanboliang commented:
@pytorchbot revert -m "This caused compilation time regression on torchbench, huggingface and dynamic models."

yanboliang commented:
@pytorchbot revert -m "This caused compilation time regression on torchbench, huggingface and dynamic models." -c nosignal

yanboliang commented:
@EikanWang This PR introduced regression of compile time for torchbench and huggingface models. I sent #106652 to revert it. Can you investigate issue and re-submit this?

The steps of benchmark compilation time:
* Run ```python benchmarks/dynamo/torchbench.py -dcuda --training --performance --inductor```
* Check the generated ```inductor.csv``` file.
* There is a column ```compilation_latency```.

Or you can run single model benchmark:
* Run ```python benchmarks/dynamo/torchbench.py -dcuda --training --performance --inductor --only hf_T5```

The following results are from my benchmarks:
```
dev,name,batch_size,speedup,abs_latency,compilation_latency,compression_ratio,eager_peak_mem,dynamo_peak_mem,calls_captured,unique_graphs,graph_breaks,unique_graph_breaks
cuda,hf_T5,8,1.229514,133.804266,48.471589,1.016064,19.291709,18.986706,795,1,8,7
cuda,hf_T5,8,1.227059,133.983976,47.352921,1.016064,19.291709,18.986706,795,1,8,7
cuda,hf_T5,8,1.234702,133.383399,54.385957,1.016064,19.291709,18.986706,795,1,8,7
cuda,hf_T5,8,1.226521,134.050623,56.280598,1.016064,19.291709,18.986706,795,1,8,7
```

The first two rows: not including this PR.
The last two rows: including this PR.

EikanWang commented:
@yanboliang , I will check the compilation latency and resubmit this PR.

EikanWang commented:
Regarding the C++/OpenMP backend,  `codecache.pick_vec_isa()` to check vectorization ISA is a time-consuming and one-shot operation. It leads to taking a longer time to import `codegen.cpp` package because the `LoopLevel` of the package is decorated by `@dataclasses.dataclass` while the decorator will invoke `codecache.pick_vec_isa()` to initialize the `simd_nelements` of the `LoopLevel`.
https://github.com/pytorch/pytorch/blob/c14cf312c9302be2cafd3be61484f5f9613fab63/torch/_inductor/codegen/cpp.py#L2883C53-L2883C53

In terms of the Triton backend, it does not need to touch it. But we'd prefer to uniform the code. Therefore, the new design registers `CpuScheduling` for CPU and `TritonScheduling` for Triton simultaneously regardless of whether the current backend is Triton. It will bring additional overhead to the Triton backend. 

```python
def init_backend_registration(self):
    if get_scheduling_for_device("cpu") is None:
        from .codegen.cpp import CppScheduling

        register_backend_for_device("cpu", CppScheduling, WrapperCodeGen)

    if get_scheduling_for_device("cuda") is None:
        from .codegen.triton import TritonScheduling

        register_backend_for_device("cuda", TritonScheduling, WrapperCodeGen)
```

To resolve the compilation latency regression for the Triton backend, we changed the `LoopLevel` a little bit by moving the `simd_nelements` to `__post_init__` and the compilation performance would be back. The new PR is https://github.com/pytorch/pytorch/pull/106874

EikanWang commented:
@yanboliang , I resubmitted this PR #106874
